### PR TITLE
Some suggestions

### PR DESCRIPTION
**Pull request by adam-mccoy**
- Move advanced options (kill on cancellation, etc) into a separate options object to clean up the API
- Remove the hooks
  - They add a lot of complexity that I don't think is warranted at this stage
  - Solve the dotnet run command problem a different way, potentially with a thin wrapper around the ShellCommand API (unfinished example included)
- Rework output targets a little bit
  - The ShellCommand API only cares about the `IOutputTarget` interface
  - That interface becomes public, allowing consumers to implement their own
  - Provide some built-in targets (string builder, callback), accessible via extension methods on ShellCommand
  - Rework the string builder target slightly to better handle newlines
- Some naming changes for consistency
  - Remove "raw" arguments - they're just arguments
  - Keep things starting with "With"

### CODE REVIEW

File: source/Tests/ShellCommandFixture.Windows.cs [hunk 1]
```diff
@@ -143,18 +150,20 @@ public async Task RunAsDifferentUser_ShouldWork(string command, string arguments
     [WindowsTheory, InlineData(SyncBehaviour.Sync), InlineData(SyncBehaviour.Async)]
     public async Task RunningAsDifferentUser_ShouldCopySpecialEnvironmentVariables(SyncBehaviour behaviour)
     {
+        var workingDirectory = Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData);
```
Comment (adam-mccoy) [inline source/Tests/ShellCommandFixture.Windows.cs:153]: I needed to add these working directories because I store my code in my user profile. When working directory is unspecified it defaults to the current directory (my user profile), which the test user does not have access to.

File: source/Shellfish/Output/StringBuilderOutputTarget.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+public class StringBuilderOutputTarget(StringBuilder stringBuilder) : IOutputTarget
+{
+    int hasWritten;
+
+    public void WriteLine(string? line)
+    {
+        if (!string.IsNullOrEmpty(line))
+        {
+            if (Interlocked.Exchange(ref hasWritten, 1) == 1)
```
Comment (borland) [inline source/Shellfish/Output/StringBuilderOutputTarget.cs:15]: Nice

File: source/Shellfish/Output/StringBuilderOutputTarget.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace Octopus.Shellfish.Output;
+
+public class StringBuilderOutputTarget(StringBuilder stringBuilder) : IOutputTarget
+{
+    int hasWritten;
+
+    public void WriteLine(string? line)
+    {
+        if (!string.IsNullOrEmpty(line))
```
Comment (borland) [inline source/Shellfish/Output/StringBuilderOutputTarget.cs:13]: We can't just ignore empty lines, otherwise if a program emits them our targets won't faithfully represent the real output

File: source/Tests/ShellCommandFixture.Windows.cs [hunk 4]
```diff
@@ -71,34 +72,27 @@ CHCP 932
...
         try
         {
-            var stdOut = new StringBuilder();
-            var stdErr = new StringBuilder();
-
-            var executor = new ShellCommand("cmd.exe")
-                .WithRawArguments("/c " + tempFile)
-                .CaptureStdOutTo(stdOut)
-                .CaptureStdErrTo(stdErr);
+            var (command1, stdOut1, stdErr1) = CreateCommand();
```
Comment (adam-mccoy) [inline source/Tests/ShellCommandFixture.Windows.cs:75]: I originally did this because the string builder target is stateful now, so it can't be reused as easily. We could use the callback target instead.

File: source/Shellfish/ShellCommand.cs [hunk 5]
```diff
@@ -49,38 +40,11 @@ public ShellCommand WithArguments(params string[] arguments)
...
-        afterExitHooks.Add(hook);
-        return this;
-    }
-
     /// <summary>
     /// Allows you to supply a string which will be passed directly to Process.StartInfo.Arguments,
     /// can be useful if you have custom quoting requirements or other special needs.
     /// </summary>
-    public ShellCommand WithRawArguments(string rawArguments)
+    public ShellCommand WithArguments(string rawArguments)
```
Comment (borland) [inline source/Shellfish/ShellCommand.cs:47]: The ergonomics on this one are tricky. Given a single argument, it's ambiguous between `WithArguments(string)` and `WithArguments(params string[])`

I'm pretty sure C# will pick this `string` one, but whether that's what the consumer intends is not clear. The quoting/escaping rules are very different, which is why I went with the "RawArguments" name to distinguish them.

Perhaps we could tweak it following what `System.Diagnostics.Process` does
- The "one big unescaped string" option is `Arguments`, so we would have `WithArguments`
- The "array of escaped strings" option is `ArgumentList`, so we would have `WithArgumentList`

?

File: source/Shellfish/ShellCommand.cs [hunk 6]
```diff
@@ -95,7 +59,7 @@ public ShellCommand WithEnvironmentVariables(Dictionary<string, string> dictiona
 #if NET5_0_OR_GREATER
     [SupportedOSPlatform("Windows")]
 #endif
-    public ShellCommand RunAsUser(NetworkCredential credential)
+    public ShellCommand WithCredentials(NetworkCredential credential)
```
Comment (borland) [inline source/Shellfish/ShellCommand.cs:62]: Fair enough :smile: Can you fix the Note: comment just below? :smile:

File: source/Tests/ShellCommandFixture.cs [hunk 7]
```diff
@@ -179,37 +180,7 @@ public async Task EchoError_ShouldWriteToCapturedStdErrCallback(SyncBehaviour be
 
         result.ExitCode.Should().Be(0, "the process should have run to completion");
         outMessages.Should().BeEmpty("no messages should be written to stdout");
-        errMessages.Should().ContainSingle(msg => msg.Contains("Something went wrong!"));
-    }
-
-    [Theory, InlineData(SyncBehaviour.Sync), InlineData(SyncBehaviour.Async)]
-    public async Task MultipleCapturingCallbacks(SyncBehaviour behaviour)
```
Comment (borland) [inline source/Tests/ShellCommandFixture.cs:186]: Why did you delete this test? It's asserting that you can call `WithOutputTarget` more than once, and the multiple targets all get called in registration order, which feels important enough to codify via a test somehow

File: source/Shellfish/DotnetCommand.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Runtime.Versioning;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Octopus.Shellfish;
+
+public class DotnetRunCommand
```
Comment (borland) [inline source/Shellfish/DotnetCommand.cs:11]: This is basically a decorator... I'm not a fan of the decorator pattern as for it to work effectively you have to implement _all_ the functionality from the decorated class, plus they're awful when debugging (i.e. most any callstack in Octopus with all the DocumentStoreDecorator classes crufting things up)

Like in this one, you haven't handled the ArgumentList case, or StdOut targets/etc.
Sure, you could add them all, but that would make it very cumbersome and it would break whenever someone added a new `WithFoo` to the ShellCommand class.

I can empathize with wanting to get rid of the hooks -- like any extensibility mechanism they do add a lot of complexity, but I don't think this is a better tradeoff.

One of the things I liked about the hook approach was that it was an escape hatch - it gave you access to mess with the `Process` object before and after it was executed, so you could do advanced things that `ShellCommand` might otherwise have not allowed. It's insurance against something like SaST deciding to ignore this whole thing because they need to set some tiny flag on the process that we don't want to (or haven't time to) expose via some public API.

I'm hopeful I can convince you there, but if you're not onboard with it, I'd be happy to drop the idea of extensibility so we can move forward. We can always add hooks/decorators later, or perhaps we'll think of a better option.

Comment (adam-mccoy) [reply source/Shellfish/DotnetCommand.cs:11]: I don't think the comparison is justified. While I can see the resemblance (if you squint), this type is under no obligation to expose the same interface as `ShellCommand`. It could offer a subset or it could be completely different. This is just what I threw together. Given that, I think pointing at the document store decorators, and their flaws, is a false equivalency.

I agree that the flexibility of hooks is a positive feature - that's the upside to the downside of the extra complexity. My argument is that this trade-off is not favourable at this stage. It may well be later.

Comment (borland) [reply source/Shellfish/DotnetCommand.cs:11]: Fair point: it looked like a decorator, so I assumed that was the pattern you were going for. I’m happy to take the hooks out and deal with the dotnet thing in the calling code until there’s some reason for us to revisit it

File: source/Shellfish/ShellCommand.cs (hunk 5 above)
Comment (borland) [reply source/Shellfish/ShellCommand.cs:47]: Determined solution:

`WithArguments(string)` for the "raw" unescaped behaviour

`WithArguments(IEnumerable<string>)` for the "ArgumentList" escaped behaviour. Deliberately no `params` as we want the choice to be intentional

File: source/Shellfish/ShellCommand.cs [hunk 9]
```diff
@@ -273,16 +217,8 @@ public ShellCommandResult Execute(CancellationToken cancellationToken = default)
...
-            if (!shouldSwallowCancellationException) throw;
-        }
-        finally
-        {
-            if (afterExitHooks is not null)
-            {
-                var exitCode = SafelyGetExitCode(process);
-                foreach (var hook in afterExitHooks) hook(exitCode, process);
-            }
+            if (options.KillProcessOnCancellation) TryKillProcessAndChildrenRecursively(process);
```
Comment (borland) [inline source/Shellfish/ShellCommand.cs:220]: Discussion: Shellfish doesn't allow you to configure this behaviour, it just does the Kill/Throw behaviour.

Let's just leave them out of the initial new API surface to keep it simple. If we find a valid use-case for them later we can add them back in.

File: source/Shellfish/Output/StringBuilderOutputTarget.cs (hunk 2 above)
Comment (borland) [reply source/Shellfish/Output/StringBuilderOutputTarget.cs:15]: Discussion: Pros and cons here. Is it better to not emit a newline on the last line, or to emit one?
There's no right answer it depends on how the caller is using it.

Decision: Let's just go with AppendLine for simplicity. We can change it later if we need to
